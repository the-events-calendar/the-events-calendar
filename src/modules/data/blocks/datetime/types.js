/**
 * Internal dependencies
 */
import { PREFIX_EVENTS_STORE } from '@moderntribe/events/data/utils';

export const SET_START_DATE_TIME = `${ PREFIX_EVENTS_STORE }/SET_START_DATE_TIME`;
export const SET_END_DATE_TIME = `${ PREFIX_EVENTS_STORE }/SET_END_DATE_TIME`;
export const SET_START_TIME = `${ PREFIX_EVENTS_STORE }/SET_START_TIME`;
export const SET_END_TIME = `${ PREFIX_EVENTS_STORE }/SET_END_TIME`;
export const SET_DATE_RANGE = `${ PREFIX_EVENTS_STORE }/SET_DATE_RANGE`;

export const SET_START_TIME_INPUT = `${ PREFIX_EVENTS_STORE }/SET_START_TIME_INPUT`;
export const SET_END_TIME_INPUT = `${ PREFIX_EVENTS_STORE }/SET_END_TIME_INPUT`;

export const SET_NATURAL_LANGUAGE_LABEL = `${ PREFIX_EVENTS_STORE }/SET_NATURAL_LANGUAGE_LABEL`;

export const SET_MULTI_DAY = `${ PREFIX_EVENTS_STORE }/SET_MULTI_DAY`;

export const SET_ALL_DAY = `${ PREFIX_EVENTS_STORE }/SET_ALL_DAY`;

export const SET_SEPARATOR_DATE = `${ PREFIX_EVENTS_STORE }/SET_SEPARATOR_DATE`;
export const SET_SEPARATOR_TIME = `${ PREFIX_EVENTS_STORE }/SET_SEPARATOR_TIME`;

export const SET_TIME_ZONE = `${ PREFIX_EVENTS_STORE }/SET_TIME_ZONE`;
export const SET_TIMEZONE_VISIBILITY = `${ PREFIX_EVENTS_STORE }/SET_TIMEZONE_VISIBILITY`;
export const SET_TIMEZONE_LABEL = `${ PREFIX_EVENTS_STORE }/SET_TIMEZONE_LABEL`;

export const SET_DATE_INPUT_VISIBILITY = `${ PREFIX_EVENTS_STORE }/SET_DATE_INPUT_VISIBILITY`;

export const SET_DATETIME_BLOCK_EDITABLE_STATE = `${ PREFIX_EVENTS_STORE }/SET_DATETIME_BLOCK_EDITABLE_STATE`; // eslint-disable-line max-len
