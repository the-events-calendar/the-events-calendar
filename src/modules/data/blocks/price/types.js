/**
 * Internal dependencies
 */
import { PREFIX_EVENTS_STORE } from '@moderntribe/events/data/utils';

export const SET_PRICE_COST = `${ PREFIX_EVENTS_STORE }/SET_PRICE_COST`;
export const SET_PRICE_SYMBOL = `${ PREFIX_EVENTS_STORE }/SET_PRICE_SYMBOL`;
export const SET_PRICE_CODE = `${ PREFIX_EVENTS_STORE }/SET_PRICE_CODE`;
export const SET_PRICE_POSITION = `${ PREFIX_EVENTS_STORE }/SET_PRICE_POSITION`;
