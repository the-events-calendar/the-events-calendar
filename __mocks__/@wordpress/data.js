/**
 * External dependencies
 */
import { noop } from 'lodash';
import { Component } from 'react';

export const select = noop;
export const withSelect = () => ( component ) => component;
export const withDispatch = () => ( component ) => component;
