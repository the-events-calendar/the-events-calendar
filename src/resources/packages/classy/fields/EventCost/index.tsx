export { default as EventCost } from './EventCost';
