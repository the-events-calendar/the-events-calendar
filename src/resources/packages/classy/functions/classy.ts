import {createRoot} from '@wordpress/element';
import {Classy} from "../elements";
import {getElement as getVisualEditorElement, toggleVisibility as toggleVisualEditorVisibility} from './visualEditor';

/**
 * Cached instance of the classy element.
 * Acts as a singleton to avoid creating multiple instances of the same element.
 *
 * @since TBD
 *
 * @type {HTMLElement|null}
 */
let classyElement: HTMLElement | null = null;

/**
 * Returns an existing Classy element or creates a new one if it doesn't exist.
 *
 * This function ensures that only one Classy element exists throughout the application
 * by caching the reference in the `classyElement` variable.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to use for element creation.
 *     If null, the global document will be used.
 * @returns {HTMLElement} The existing or newly created Classy element.
 */
export function getOrCreateElement(document: Document | null = null): HTMLElement {
	document = document ?? window.document;

	if (classyElement !== null) {
		return classyElement;
	}

	const element = document.createElement('div');
	element.id = 'tec-classy';
	element.classList.add('classy-root', 'classy-root--admin');
	classyElement = element;

	return element;
}

/**
 * Inserts the Classy element into the DOM after the visual editor.
 *
 * This function will get or create the Classy element and insert it into the DOM
 * immediately after the visual editor element. If the visual editor cannot be found,
 * the function will return false indicating failure.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to operate within.
 *     If null, the global document will be used.
 *
 * @returns {boolean} True if insertion was successful, false otherwise.
 */
export function insertElement(document: Document | null = null): boolean {
	document = document ?? window.document;
	const element = getOrCreateElement(document);
	const visualEditor = getVisualEditorElement(document);

	if (visualEditor === null) {
		return false;
	}

	visualEditor.parentNode.insertBefore(element, visualEditor.nextSibling);
	toggleVisualEditorVisibility(document);

	return true;
}

/**
 * Toggles the visibility of the Classy element.
 *
 * This function adds or removes the 'classy-root--hidden' class from the Classy element,
 * which should be styled with CSS to show or hide the element accordingly.
 * The element will be created if it doesn't already exist.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to operate within.
 *     If null, the global document will be used.
 *
 * @returns {void}
 */
export function toggleElementVisibility(document: Document | null = null): void {
	document = document ?? window.document;
	getOrCreateElement().classList.toggle('classy-root--hidden');
	toggleVisualEditorVisibility(document);
}

/**
 * Initializes the application by creating and rendering the Classy root element.
 *
 * This function ensures that the Classy element is either retrieved or created
 * and then initializes the React application by rendering the `Classy` component
 * into that element using React's `createRoot` API.
 *
 * @since TBD
 *
 * @param {Document|null} document - The document to use for element creation.
 *	 If null, the global document will be used.
 *
 * @returns {void}
 */
export function initApp(document: Document | null = null): void {
	document = document ?? window.document;
	const classyRoot = createRoot(getOrCreateElement(document));
	classyRoot.render(Classy());
}
