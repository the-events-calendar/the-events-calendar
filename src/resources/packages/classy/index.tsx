import {dispatch, select, subscribe} from "@wordpress/data";
import {localizedData} from './localized-data';
import {RichText} from '@wordpress/block-editor';
import {__} from "@wordpress/i18n";
import {registerPlugin} from '@wordpress/plugins';
import {createRoot} from '@wordpress/element';
import { SlotFillProvider, Slot, Fill } from '@wordpress/components';
import { addFilter, applyFilters } from '@wordpress/hooks';
import './style.pcss';

console.log('Classy!');

const {eventCategoryTaxonomyName} = localizedData;

function whenEditorIsReady(): Promise<void> {
	return new Promise((resolve: Function) => {
		const unsubscribe: Function = subscribe(() => {
			// This will trigger after the initial render blocking, before the window load event
			// This seems currently more reliable than using __unstableIsEditorReady
			if (select('core/editor').isCleanNewPost() || select('core/block-editor').getBlockCount() > 0) {
				unsubscribe()
				resolve()
			}
		})
	})
}

function createClassyElement(): Element {
	const classy = document.createElement('div');
	classy.id = 'tec-classy';
	classy.classList.add('classy-root', 'classy-root--admin');

	return classy;
}

function injectClassyElement(classyElement: Element): void {
	const visualEditor = document.querySelector('.editor-visual-editor');

	if (!visualEditor) {
		return;
	}

	visualEditor.parentNode.insertBefore(classyElement, visualEditor.nextSibling);
}

// Remove the category sidebar element.
dispatch('core/editor').removeEditorPanel(`taxonomy-panel-${eventCategoryTaxonomyName}`);

// Remove the post tag sidebar element.
dispatch('core/editor').removeEditorPanel('taxonomy-panel-post_tag');


function hideZoomOutButton(): void {
	// Remove the Zoom Out button. The only way is by its aria label.
	const zoomOutAriaLabel = __('Zoom Out');
	document.querySelectorAll(`.components-button[aria-label="${zoomOutAriaLabel}"]`)
		.forEach((el: Element) => el.style.display = 'none');
}

function hideInserterToggle() {
	// Remove the inserter toggle button.
	document.querySelectorAll('.editor-document-tools__inserter-toggle')
		.forEach((button: Element) => button.style.display = 'none');
}

function addEditorTools(onClick: (this: GlobalEventHandlers, ev: MouseEvent) => void): void {
	let editorToolsAdded = false;

	function EditorTools(): void {
		if (editorToolsAdded) {
			return;
		}

		const editorDocumentTools = document.querySelector('.editor-document-tools .editor-document-tools__left')

		if (editorDocumentTools) {
			const previewButton = document.createElement('button')
			previewButton.classList.add('tec-editor-tool', 'tec-editor-tool--preview', 'button');
			previewButton.type = 'button';
			previewButton.dataset.toolbarItem = 'true';
			previewButton.innerHTML = `<span class="dashicons dashicons-visibility"></span> ${__('Visual', 'the-events-calendar')}`;
			editorDocumentTools.append(previewButton);
			editorToolsAdded = true;
			previewButton.onclick = onClick;
		}

		return null;
	}

	registerPlugin('tec-editor-tools', {
		render: EditorTools
	});
}

const classyElement = createClassyElement();

whenEditorIsReady().then(() => {
	hideZoomOutButton();
	hideInserterToggle();

	addFilter(
		'classy.render',
		'additional-plugin',
		()=>(
			<Fill name="classy.fields">
				<div className="classy-input">
					<div className="classy-input__label">
						Fill
					</div>
					<div className="classy-input__input">
						<input type="text" placeholder="If you can read this, fill works!"/>
					</div>
				</div>
			</Fill>
		)
	);

	const classyRoot = createRoot(classyElement);

	classyRoot.render(
		<SlotFillProvider>

			{applyFilters('classy.render')}

			<div className="classy-container">
				<h3 className="classy-title">New Event</h3>

				<div className="classy-input">
					<div className="classy-input__label">
						Title
					</div>
					<div className="classy-input__input">
						<input type="text" placeholder="Your title"/>
					</div>
				</div>

				<div className="classy-input">
					<div className="classy-input__label">
						Description
					</div>
					<div className="classy-input__input">
						<RichText
							tagName="p"
							value=""
							allowedFormats={['core/bold', 'core/italic']}
							placeholder="Your event description"
						/>
					</div>
				</div>

				<Slot name="classy.fields" />

			</div>

		</SlotFillProvider>
	);

	injectClassyElement(classyElement);
	const toggleClassyVisibility = (): void => {
		console.log('classyElement', classyElement);
		classyElement.classList.toggle('classy-root--hidden');
	};
	addEditorTools(toggleClassyVisibility);
});
