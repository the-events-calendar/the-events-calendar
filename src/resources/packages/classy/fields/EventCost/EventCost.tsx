import * as React from 'react';
import { Fragment, useEffect, useState } from 'react';
import { __, _x } from '@wordpress/i18n';
import { __experimentalInputControl as InputControl, ToggleControl } from '@wordpress/components';
import { useDispatch, useSelect } from '@wordpress/data';
import { CurrencySelector } from '@tec/common/classy/components';
import {
	METADATA_EVENT_COST,
	METADATA_EVENT_CURRENCY,
	METADATA_EVENT_CURRENCY_POSITION,
	METADATA_EVENT_CURRENCY_SYMBOL,
	METADATA_EVENT_IS_FREE
} from '../../constants';
import { Currency } from '@tec/common/classy/types/Currency';
import { CurrencyPosition } from '@tec/common/classy/types/CurrencyPosition';

export default function EventCost(): JSX.Element {
	const { meta, defaultCurrency } = useSelect( ( select ) => {
		const { getEditedPostAttribute }: {
			getEditedPostAttribute: ( attribute: string ) => any
		} = select( 'core/editor' );
		const { getDefaultCurrency }: {
			getDefaultCurrency: () => Currency
		} = select( 'tec/classy' );
		return {
			meta: getEditedPostAttribute( 'meta' ) || {},
			defaultCurrency: getDefaultCurrency(),
		};
	}, [] );

	const freeText = __( 'Free', 'the-events-calendar' );

	const { editPost } = useDispatch( 'core/editor' );

	const isFreeMeta: boolean = meta[ METADATA_EVENT_IS_FREE ] || false;
	const [ isFree, setIsFree ] = useState< boolean >( isFreeMeta );

	const eventCostMeta: string = meta[ METADATA_EVENT_COST ] || '';
	const [ eventCostValue, setEventCostValue ] = useState< string >( isFree ? freeText : eventCostMeta );

	const eventCurrencySymbolMeta: string = meta[ METADATA_EVENT_CURRENCY_SYMBOL ] || defaultCurrency.symbol;
	const [ currencySymbol, setCurrencySymbol ] = useState< string >( eventCurrencySymbolMeta );

	const eventCurrencyPosition: CurrencyPosition =
		meta[ METADATA_EVENT_CURRENCY_POSITION ] ||
		defaultCurrency.position;
	const [ currencyPosition, setCurrencyPosition ] = useState< CurrencyPosition >( eventCurrencyPosition );

	const [ costHasFocus, setCostHasFocus ] = useState< boolean >( false );

	// Track changes to the post content and update the state accordingly.
	useEffect( () => {
		setEventCostValue( eventCostMeta );
	}, [ eventCostMeta ] );

	useEffect( () => {
		setIsFree( isFreeMeta );
	}, [ isFreeMeta ] );

	useEffect( () => {
		setCurrencySymbol( eventCurrencySymbolMeta );
	}, [ eventCurrencySymbolMeta ] );

	useEffect( () => {
		setCurrencyPosition( eventCurrencyPosition );
	}, [ eventCurrencyPosition ] );

	// Handle changes to the event cost input.
	const onCostChange = ( nextValue: string | undefined ): void => {
		setEventCostValue( nextValue ?? '' );
		editPost( { meta: { [ METADATA_EVENT_COST ]: nextValue } } );
	};

	// Handle changes to the "is free" toggle.
	const onFreeChange = ( nextValue: boolean ): void => {
		setIsFree( nextValue );
		setEventCostValue( nextValue ? freeText : eventCostMeta );

		editPost( { meta: { [ METADATA_EVENT_IS_FREE ]: nextValue } } );
	};

	/**
	 * Formats a currency value based on the current currency position.
	 *
	 * This function takes a numeric value as a string and formats it with the currency symbol,
	 * according to the specified currency position (prefix or postfix).
	 *
	 * @since TBD
	 *
	 * @param {string} value
	 * @return {string} The formatted currency value.
	 */
	const formatCurrencyValue = ( value: string ): string => {
		return currencyPosition === 'prefix'
			? `${ currencySymbol }${ value }`
			: `${ value }${ currencySymbol }`;
	}

	/**
	 * Formats the event cost value for display.
	 *
	 * This function formats the event cost value based on whether the input has focus or if the event is free.
	 * It handles multiple prices separated by commas and returns a formatted string.
	 *
	 * @since TBD
	 *
	 * @param {string} value The raw event cost value as a string.
	 * @return {string} The formatted event cost value.
	 */
	const formatEventCostValue = ( value: string ): string => {
		// If the cost input has focus or the event is free, return the value as is.
		if ( costHasFocus || isFree ) {
			return value;
		}

		const pieces = value
			.split( ',' )
			.map( ( piece ) => piece.trim() )
			.filter( ( piece ) => piece !== '' );

		if ( pieces.length === 0 ) {
			return value;
		}

		// Convert pieces to numbers and find min/max
		const numbers = pieces.map( piece => {
			const num = parseFloat( piece );
			return isNaN( num ) ? 0 : Number( num.toFixed( 2 ) );
		} );

		const min = Math.min( ...numbers );
		const max = Math.max( ...numbers );
		const formattedMin = formatCurrencyValue( min.toFixed( 2 ) );
		const formattedMax = formatCurrencyValue( max.toFixed( 2 ) );

		// If min and max are the same, return just one formatted number
		if ( min === max ) {
			return formatCurrencyValue( min.toFixed( 2 ) );
		}

		// Otherwise return the range
		return `${ formattedMin } - ${ formattedMax }`;
	};

	return (
		<Fragment>
			<div className="classy-field__group classy-field__event-cost">
				<div className="classy-field__input classy-field__input-full-width">
					<div className="classy-field__control classy-field__control--input">
						<InputControl
							label={
								<span className="classy-field__input-title">
									{ _x( 'Event cost', 'Event cost input title', 'the-events-calendar' ) }
								</span>
							}
							value={ formatEventCostValue( eventCostValue ) }
							onChange={ onCostChange }
							disabled={ isFree }
							onFocus={ () => setCostHasFocus( true ) }
							onBlur={ () => setCostHasFocus( false ) }
						/>
					</div>
				</div>

				<div className="classy-field__input classy-field__input--height-100">
					<CurrencySelector
						currencyCodeMeta={ METADATA_EVENT_CURRENCY }
						currencyPositionMeta={ METADATA_EVENT_CURRENCY_POSITION }
						currencySymbolMeta={ METADATA_EVENT_CURRENCY_SYMBOL }
					/>
				</div>

				<div className="classy-field__input classy-field__input--height-75">
					<ToggleControl
						label={ _x( 'Event is free', 'Event cost toggle label', 'the-events-calendar' ) }
						checked={ isFree }
						onChange={ onFreeChange }
						__nextHasNoMarginBottom
					/>
				</div>
			</div>

			<div className="classy-field__input-note">
				{ _x(
					'If multiple entry prices are available, list each price separated by commas.',
					'Event cost input note',
					'the-events-calendar'
				) }
			</div>
		</Fragment>
	);
}
