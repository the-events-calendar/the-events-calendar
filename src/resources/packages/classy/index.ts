import { addAction, addFilter, didAction, doAction } from '@wordpress/hooks';
import renderFields from './functions/renderFields';
import { storeConfig } from './store';
import { STORE_NAME } from './constants';
import { getRegistry } from '@tec/common/classy/store';
import * as constants from './constants';
import { getNewStartEndDates, getAllDayNewDates, getMultiDayDates } from './functions/events';

/**
 * Hook on the Classy application initialization to add TEC store to the Classy registry.
 *
 * @since TBD
 *
 * @return {void} The ECP store is registered.
 */
const registerStore = (): void => {
	getRegistry().registerStore( STORE_NAME, storeConfig );

	/**
	 * Fires after the TEC store is registered and the TEC Classy application is initialized.
	 *
	 * @since TBD
	 *
	 * @return {void} The TEC store is registered.
	 */
	doAction( 'tec.classy.events.initialized' );
};

if ( didAction( 'tec.classy.initialized' ) ) {
	registerStore();
} else {
	addAction( 'tec.classy.initialized', 'tec.classy.events', registerStore );
}

// Hook on the Classy fields rendering logic to render the fields.
addFilter( 'tec.classy.render', 'tec.classy.events', renderFields );

// Export constants and functions to allow their reference in other packages.
export {
	constants,
	getNewStartEndDates,
	getAllDayNewDates,
	getMultiDayDates,
};
