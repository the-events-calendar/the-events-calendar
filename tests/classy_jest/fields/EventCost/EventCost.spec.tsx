import * as React from 'react';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { beforeEach, describe, expect, it, jest } from '@jest/globals';
import { RegistryProvider } from '@wordpress/data';
import { createRegistry } from '../../../../common/src/resources/packages/classy/store';
import EventCost from '../../../../src/resources/packages/classy/fields/EventCost/EventCost';
import { Currency } from '../../../../common/src/resources/packages/classy/types/Currency';
import { CurrencyPosition } from '../../../../common/src/resources/packages/classy/types/CurrencyPosition';

type Selector = {
	'core/editor': {
		getEditedPostAttribute: ( attribute: string ) => any;
	};
	'tec/classy': {
		getDefaultCurrency: () => Currency;
	};
};

// Mock WordPress data store
jest.mock( '@wordpress/data', () => {
	const mockSelect = ( storeName: string ) => {
		if ( storeName === 'core/editor' ) {
			return {
				getEditedPostAttribute: () => ( {
					meta: {
						tribe_event_cost: '',
						tribe_event_currency_symbol: '$',
						tribe_event_currency_position: 'prefix',
						tribe_event_is_free: false,
					},
				} ),
			};
		}
		if ( storeName === 'tec/classy' ) {
			return {
				getDefaultCurrency: () => ( {
					code: 'USD',
					symbol: '$',
					position: 'prefix',
				} ),
			};
		}
		return {};
	};

	const mockDispatch = ( storeName: string ) => {
		if ( storeName === 'core/editor' ) {
			return {
				editPost: jest.fn(),
			};
		}
		return {};
	};

	return {
		useSelect: ( callback: ( select: typeof mockSelect ) => any ) => callback( mockSelect ),
		useDispatch: ( storeName: string ) => mockDispatch( storeName ),
		RegistryProvider: ( { children } ) => children,
	};
} );

// Mock WordPress components
jest.mock( '@wordpress/components', () => ( {
	__experimentalInputControl: ( { label, value, onChange, disabled, onFocus, onBlur } ) => (
		<input
			aria-label={ label.props.children }
			value={ value }
			onChange={ ( e ) => onChange( e.target.value ) }
			disabled={ disabled }
			onFocus={ onFocus }
			onBlur={ onBlur }
		/>
	),
	ToggleControl: ( { label, checked, onChange } ) => (
		<label>
			{ label }
			<input type="checkbox" checked={ checked } onChange={ ( e ) => onChange( e.target.checked ) } />
		</label>
	),
} ) );

// Mock CurrencySelector component
jest.mock( '../../../../common/src/resources/packages/classy/components', () => ( {
	CurrencySelector: () => <div className="classy-field__currency-selector" />,
} ) );

describe( 'EventCost Component', () => {
	const mockDefaultCurrency: Currency = {
		code: 'USD',
		symbol: '$',
		position: 'prefix' as CurrencyPosition,
	};

	const mockMeta = {
		[ 'tribe_event_cost' ]: '',
		[ 'tribe_event_currency_symbol' ]: '$',
		[ 'tribe_event_currency_position' ]: 'prefix',
		[ 'tribe_event_is_free' ]: false,
	};

	beforeEach( () => {
		jest.resetAllMocks();
	} );

	it( 'renders correctly with default props', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		expect( container ).toMatchSnapshot();
	} );

	it( 'displays "Free" when event is free', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		const freeToggle = screen.getByLabelText( 'Event is free' );
		await userEvent.click( freeToggle );

		expect( container ).toMatchSnapshot();
	} );

	it( 'formats single cost value correctly', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		const costInput = screen.getByLabelText( 'Event cost' );
		await userEvent.type( costInput, '10.50' );

		expect( container ).toMatchSnapshot();
	} );

	it( 'formats multiple cost values correctly', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		const costInput = screen.getByLabelText( 'Event cost' );
		await userEvent.type( costInput, '10.50, 20.75, 15.25' );

		expect( container ).toMatchSnapshot();
	} );

	it( 'handles invalid cost values gracefully', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		const costInput = screen.getByLabelText( 'Event cost' );
		await userEvent.type( costInput, 'invalid, 20.75' );

		expect( container ).toMatchSnapshot();
	} );

	it( 'updates currency symbol and position correctly', async () => {
		const registry = await createRegistry();
		const { container } = render(
			<RegistryProvider value={ registry }>
				<EventCost />
			</RegistryProvider>
		);

		const costInput = screen.getByLabelText( 'Event cost' );
		await userEvent.type( costInput, '10.50' );

		// Simulate currency change
		const currencySelector = container.querySelector( '.classy-field__currency-selector' );
		expect( currencySelector ).not.toBeNull();

		expect( container ).toMatchSnapshot();
	} );
} );
