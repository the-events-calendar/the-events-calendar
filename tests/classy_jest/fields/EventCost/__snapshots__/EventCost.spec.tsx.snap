// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCost Component displays "Free" when event is free 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          disabled=""
          value="Free"
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;

exports[`EventCost Component formats multiple cost values correctly 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          value="10.50, 20.75, 15.25"
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;

exports[`EventCost Component formats single cost value correctly 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          value="10.50"
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;

exports[`EventCost Component handles invalid cost values gracefully 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          value="invalid, 20.75"
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;

exports[`EventCost Component renders correctly with default props 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          value=""
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;

exports[`EventCost Component updates currency symbol and position correctly 1`] = `
<div>
  <div
    class="classy-field__group classy-field__event-cost"
  >
    <div
      class="classy-field__input classy-field__input-full-width"
    >
      <div
        class="classy-field__control classy-field__control--input"
      >
        <input
          aria-label="Event cost"
          value="10.50"
        />
      </div>
    </div>
    <div
      class="classy-field__input classy-field__input--height-100"
    >
      <div
        class="classy-field__currency-selector"
      />
    </div>
    <div
      class="classy-field__input classy-field__input--height-75"
    >
      <label>
        Event is free
        <input
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    class="classy-field__input-note"
  >
    If multiple entry prices are available, list each price separated by commas.
  </div>
</div>
`;
